File: rtls-kmp/rtls-offline-sync/src/commonMain/kotlin/com/rtls/sync/OfflineSyncClient.kt
package com.rtls.sync

import com.rtls.core.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.SharedFlow

/**
 * Standalone offline-first sync client. Usable without location or WebSocket packages.
 * Insert data from any source; SyncEngine handles batch upload and optional bidirectional pull.
 */
class OfflineSyncClient(
    private val store: LocationStore,
    private val api: LocationSyncAPI,
    private val batchingPolicy: BatchingPolicy = BatchingPolicy(),
    private val retryPolicy: SyncRetryPolicy = SyncRetryPolicy.Default,
    private val retentionPolicy: RetentionPolicy = RetentionPolicy.Recommended,
    private val networkMonitor: NetworkMonitor? = null,
    private val pullAPI: LocationPullAPI? = null,
    private val mergeStrategy: LocationMergeStrategy? = null,
    private val pullIntervalSeconds: Long? = null,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
) {
    private val engine = SyncEngine(
        store = store,
        api = api,
        batching = batchingPolicy,
        retryPolicy = retryPolicy,
        retentionPolicy = retentionPolicy,
        networkMonitor = networkMonitor,
        scope = scope,
        pullAPI = pullAPI,
        mergeStrategy = mergeStrategy,
        pullIntervalSeconds = pullIntervalSeconds
    )

    val events: SharedFlow<SyncEngineEvent> = engine.events

    fun start() = engine.start()
    fun stop() = engine.stop()

    suspend fun insert(points: List<LocationPoint>) {
        store.insert(points)
        engine.notifyNewData()
    }

    suspend fun flushNow(maxBatches: Int? = null) = engine.flushNow(maxBatches)
    suspend fun pullNow() = engine.pullNow()

    suspend fun pendingCount(): Int = store.pendingCount()
    suspend fun pendingStats(): PendingStats = store.pendingStats()
}

File: Sources/RTLSOfflineSync/OfflineSyncClient.swift
import Foundation
import RTLSCore

/// Standalone offline-first sync client. Usable without location or WebSocket packages.
/// Insert data from any source; SyncEngine handles batch upload and optional bidirectional pull.
public actor OfflineSyncClient {
    private let store: SQLiteLocationStore
    private let syncEngine: SyncEngine

    private let eventsStream: AsyncStream<SyncEngineEvent>
    private let continuation: AsyncStream<SyncEngineEvent>.Continuation

    public nonisolated var events: AsyncStream<SyncEngineEvent> { eventsStream }

    public init(
        store: SQLiteLocationStore,
        api: any LocationSyncAPI,
        batchingPolicy: BatchingPolicy = BatchingPolicy(),
        retryPolicy: SyncRetryPolicy = .default,
        retentionPolicy: RetentionPolicy = .recommended,
        network: NetworkMonitor = NetworkMonitor(),
        pullAPI: (any LocationPullAPI)? = nil,
        mergeStrategy: (any LocationMergeStrategy)? = nil,
        pullInterval: TimeInterval? = nil
    ) {
        self.store = store
        self.syncEngine = SyncEngine(
            store: store,
            api: api,
            batchingPolicy: batchingPolicy,
            retryPolicy: retryPolicy,
            retentionPolicy: retentionPolicy,
            network: network,
            pullAPI: pullAPI,
            mergeStrategy: mergeStrategy,
            pullInterval: pullInterval
        )

        let (stream, cont) = AsyncStream<SyncEngineEvent>.makeStream(bufferingPolicy: .bufferingNewest(64))
        self.eventsStream = stream
        self.continuation = cont

        Task { [syncEngine] in
            for await event in syncEngine.events {
                cont.yield(event)
            }
        }
    }

    public func start() async {
        await syncEngine.start()
    }

    public func stop() async {
        await syncEngine.stop()
    }

    /// Insert points from any data source. SyncEngine will batch-upload them.
    public func insert(points: [LocationPoint]) async throws {
        try await store.insert(points: points)
        await syncEngine.notifyNewData()
    }

    public func flushNow(maxBatches: Int? = nil) async {
        await syncEngine.flushNow(maxBatches: maxBatches)
    }

    public func pullNow() async {
        await syncEngine.pullNow()
    }

    public func pendingCount() async throws -> Int {
        try await store.pendingCount()
    }

    public func pendingStats() async throws -> PendingStats {
        try await store.pendingStats()
    }
}
